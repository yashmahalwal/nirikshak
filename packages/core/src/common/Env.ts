/* istanbul ignore file */
/* eslint-disable prefer-const */
export let MAX_PLURAL_ENTRIES = 10;
export let MIN_PLURAL_ENTRIES = 0;
export let RANDOMNESS_ITERATIONS = 10;
