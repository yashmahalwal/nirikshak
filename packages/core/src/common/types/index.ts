export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
