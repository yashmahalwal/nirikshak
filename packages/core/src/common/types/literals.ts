/* Literals that can be provided to describe the resource */

// Basic literal type
// Ex: 1, false, null, "myString"
export type Literal = string | number | null | boolean;

// type guard: Literal
export function isLiteral(input: any): input is Literal {
    const c = typeof input;
    switch (c) {
        case "number":
        case "string":
        case "boolean":
            return true;
        default:
            return input === null;
    }
}
