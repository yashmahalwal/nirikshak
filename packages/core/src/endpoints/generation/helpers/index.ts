export { generateBodyType } from "./bodyTypeGen";
export { generateHeaderAndStatus } from "../output/headerAndStatus";
