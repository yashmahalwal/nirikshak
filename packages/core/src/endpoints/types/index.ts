export {
    MethodType,
    HeaderMap,
    Cases,
    BodyType,
    BodyInstance,
} from "./helpers";
export { isURLString } from "./urlString";
export { isResourceString } from "./resourceString";
export { isDescription } from "./description";
export { HeaderAndQuery, Inputs } from "./input";
export { HeaderAndStatus, Outputs } from "./output";
