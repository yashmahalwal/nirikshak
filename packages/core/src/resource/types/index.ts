export { isResource } from "./resource";
export { Resource } from "./resource";
export { ResourceInstance } from "./helper";
