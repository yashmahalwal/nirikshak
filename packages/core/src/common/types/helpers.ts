import { CustomFunction } from "./custom";
import { Literal, isLiteral } from "./literals";

export interface SchemaHelpers {
    [key: string]: CustomFunction;
}

// Ultimately, a resource fields ends in Primitive type
// Ex: {age: 23}
// Ex: {address: [{zipCode: "370445"}, {zipCode: "121342"}], classes: [1,2,3,4,5]}
export type Primitives = Literal | Array<Primitives>;
export function isPrimitives(input: any): input is Primitives {
    if (Array.isArray(input)) return input.every((i) => isPrimitives(i));
    return isLiteral(input);
}
