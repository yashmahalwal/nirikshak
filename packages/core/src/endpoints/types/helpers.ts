import { ResourceString, isResourceString } from "./resourceString";
import { FakerType, isFakerType } from "../../common/types/fakerTypes";
import {
    CustomFunctionType,
    isCustomFunction,
} from "../../common/types/custom";
import { Literal, isLiteral } from "../../common/types/literals";
import { Primitives } from "../../common/types/helpers";

export type MethodType = "DELETE" | "GET" | "PUT" | "POST" | "PATCH";
export function isMethodType(key: any): key is MethodType {
    if (typeof key !== "string") return false;

    switch (key) {
        case "DELETE":
        case "GET":
        case "PUT":
        case "POST":
        case "PATCH":
            return true;
        default:
            return false;
    }
}

export type Cases = "NEGATIVE" | "POSITIVE" | "DESTRUCTIVE";

export function isCaseType(key: any): key is Cases {
    if (typeof key !== "string") return false;
    switch (key) {
        case "NEGATIVE":
        case "POSITIVE":
        case "DESTRUCTIVE":
            return true;
        default:
            return false;
    }
}

export type BaseType =
    | FakerType
    | CustomFunctionType
    | ResourceString
    | Literal;

export function isBaseType(input: any): input is BaseType {
    return (
        isFakerType(input) ||
        isCustomFunction(input) ||
        isResourceString(input) ||
        isLiteral(input)
    );
}

export type Base = BaseType | Base[];

export function isBase(input: any): input is Base {
    if (Array.isArray(input)) return input.every((value) => isBase(value));
    return isBaseType(input);
}

export type WithModifiers<T extends Base | BodyType> = (T extends Base
    ? { type: Base }
    : { field: BodyType }) & {
    plural?: boolean;
    nullable?: boolean;
    optional?: boolean;
};

// Checking the validity of the modifier fields of the HOT
function checkModifiers(input: WithModifiers<Base | BodyType>): boolean {
    let value = true;
    void (
        "nullable" in input &&
        (value = value && typeof input["nullable"] === "boolean")
    );
    void (
        "optional" in input &&
        (value = value && typeof input["optional"] === "boolean")
    );
    void (
        "plural" in input &&
        (value = value && typeof input["plural"] === "boolean")
    );

    return value;
}

// type guard: WithModifiers<ResourceBase>
export function isWithModifiersBodyType(
    input: any
): input is WithModifiers<BodyType> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("field" in input) {
        return isBodyType(input["field"]);
    }
    return false;
}

// type guard: WithModifies<BaseType>
export function isWithModifiersBase(input: any): input is WithModifiers<Base> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("type" in input) {
        return isBase(input["type"]);
    }
    return false;
}

// type guard: WithModifiers<BaseType> or WithModifiers<ResourceBase>
// Infers from input
export function isWithModifiers(
    input: any
): input is WithModifiers<Base | BodyType> {
    return isWithModifiersBase(input) || isWithModifiersBodyType(input);
}

export type OneOfEntries =
    | { fields: (BodyType | WithModifiers<BodyType>)[] }
    | { types: (Base | WithModifiers<Base>)[] }
    | {
          fields: (BodyType | WithModifiers<BodyType>)[];
          types: (Base | WithModifiers<Base>)[];
      };

// type guard: OneOfEntries
export function isOneOfEntries(input: any): input is OneOfEntries {
    if (!input || typeof input !== "object") return false;
    if (!Array.isArray(input["fields"]) && !Array.isArray(input["types"]))
        return false;

    if (Array.isArray(input["fields"]) && !input["fields"].length) return false;
    if (Array.isArray(input["types"]) && !input["types"].length) return false;

    let value = true;
    value =
        value &&
        (Array.isArray(input["types"])
            ? input["types"].every(
                  (entry) => isBase(entry) || isWithModifiersBase(entry)
              )
            : true);

    value =
        value &&
        (Array.isArray(input["fields"])
            ? input["fields"].every(
                  (entry) => isBodyType(entry) || isWithModifiersBodyType(entry)
              )
            : true);

    return value;
}

export interface BodyType {
    [key: string]:
        | ResourceString
        | Exclude<Base, ResourceString>
        | WithModifiers<Base>
        | WithModifiers<BodyType>
        | OneOfEntries
        | BodyType;
}

export interface BodyInstance {
    [key: string]: Primitives | BodyInstance | BodyInstance[];
}

export function isBodyType(input: any): input is BodyType {
    if (!input || typeof input !== "object" || Array.isArray(input))
        return false;

    return Object.values(input).every(
        (entry) =>
            isResourceString(entry) ||
            isBase(entry) ||
            isWithModifiers(entry) ||
            isOneOfEntries(entry) ||
            isBodyType(entry)
    );
}

export interface HeaderMap {
    [key: string]: Exclude<BaseType, null>;
}

export function isHeaderMap(input: any): input is HeaderMap {
    if (!input || typeof input !== "object" || Array.isArray(input))
        return false;

    return Object.values(input).every(
        (value) => value !== null && isBaseType(value)
    );
}
