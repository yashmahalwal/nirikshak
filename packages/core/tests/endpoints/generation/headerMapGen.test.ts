import faker from "faker";
import { SchemaHelpers } from "../../../src/common/types/helpers";
import { HeaderMap } from "../../../src/endpoints/types/helpers";
import {
    HeadersInstance,
    generateHeaderMap,
} from "../../../src/endpoints/generation/helpers/headerMapGen";
import { RANDOMNESS_ITERATIONS } from "../../../src/common/Env";
import { Literal } from "../../../src/common/types/literals";
const resource = {
    id: faker.random.uuid(),
    name: {
        firstName: faker.name.firstName(),
        lastName: faker.name.lastName(),
    },
    addresses: [
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
    ],
    recursion: {
        bases: {
            args: {
                plethora: [1, 2, 10, [[7, 5, [4]]], 3],
            },
        },
    },
};

const Helpers: SchemaHelpers = {
    word: (): Promise<string> =>
        new Promise((resolve) =>
            setTimeout(() => resolve(faker.random.word()), 10)
        ),
    async invalidHelper() {
        return { age: 10 } as any;
    },
};

const ValidHeaders: {
    input: HeaderMap;
    output: () => Promise<HeadersInstance>;
}[] = [
    {
        input: {
            entry1: "1000",
            entry2: "resource:id",
            entry3: "resource:addresses[0].zipCode",
            words: "custom:word",
            number1: "faker:random.number",
        },
        async output(): Promise<HeadersInstance> {
            const headers: HeadersInstance = {};
            const promiseMap: Map<string, Promise<Literal>> = new Map();

            for (const key in this.input) {
                switch (key) {
                    case "entry1":
                        promiseMap.set(key, Promise.resolve("1000"));
                        break;
                    case "entry2":
                        promiseMap.set(key, Promise.resolve(resource.id));
                        break;
                    case "entry3":
                        promiseMap.set(
                            key,
                            Promise.resolve(resource.addresses[0].zipCode)
                        );
                        break;
                    case "words":
                        promiseMap.set(key, Helpers.word());
                        break;
                    case "number1":
                        promiseMap.set(
                            key,
                            Promise.resolve(faker.random.number())
                        );
                        break;
                }
            }

            for (const key in this.input)
                headers[key] = (await promiseMap.get(key)) as Exclude<
                    Literal,
                    null
                >;

            return headers;
        },
    },
];

const InvalidHeaders: any[] = [
    { entry1: null },
    { entry2: "resource:addresses" },
    { entry4: "custom:nonExistent" },
    { entry5: "resource:lorem.ipsum" },
    { entry6: "custom:invalidHelper" },
];

describe("Header map generation", () => {
    for (let i = 1; i <= RANDOMNESS_ITERATIONS; i++)
        ValidHeaders.forEach((entry, index) => {
            test(`Valid header entry: ${index}, iteration: ${i}`, async () => {
                faker.seed(123 * i + index);
                const got = await generateHeaderMap(
                    entry.input,
                    resource,
                    Helpers
                );
                faker.seed(123 * i + index);
                const expected = await entry.output();
                expect(got).toEqual(expected);
            });
        });

    test.each(InvalidHeaders)(`Invalid header entry: %#`, (entry) =>
        expect(
            generateHeaderMap(entry, resource, Helpers)
        ).rejects.toMatchSnapshot()
    );
});
