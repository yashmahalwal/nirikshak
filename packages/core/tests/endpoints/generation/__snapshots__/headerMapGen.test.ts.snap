// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header map generation Invalid header entry: 0 1`] = `[Error: Header: entry1 = null resolved to null value]`;

exports[`Header map generation Invalid header entry: 1 1`] = `[Error: Resource string resource:addresses resolved to non primitive value]`;

exports[`Header map generation Invalid header entry: 2 1`] = `[Error: function: custom:nonExistent, args: ]`;

exports[`Header map generation Invalid header entry: 3 1`] = `[Error: Invalid resource instance path: lorem.ipsum]`;

exports[`Header map generation Invalid header entry: 4 1`] = `[Error: Non literal type value from custom:invalidHelper]`;
